Updates DNSSEC07 test case specification (#1425)
dotse/zonemaster: docs/public/specifications/tests/DNSSEC-TP/dnssec07.md

# DNSSEC07: DNSSEC signed zone and DS in parent for signed zone

## Test case identifier
**DNSSEC07**

## Table of contents

* [Objective](#objective)
* [Scope](#scope)
* [Inputs](#inputs)
* [Summary](#summary)
  * [Restrictions on combinations of the message tags](#restrictions-on-combinations-of-the-message-tags)
* [Test procedure]
* [Outcome(s)](#outcomes)
* [Special procedural requirements](#special-procedural-requirements)
* [Intercase dependencies](#intercase-dependencies)
* [Terminology](#terminology)

## Objective

DNSSEC is the security upgrade of DNS, in the same way as TLS is the security

> upgrade of HTTP. If a zone is upgrades to DNSSEC it means that it is signed by

Author: I agree. I will rewrite that sentence.

# DNSSEC07: DNSSEC signed zone and DS in parent for signed zone

## Test case identifier
**DNSSEC07**

## Table of contents

* [Objective](#objective)
* [Scope](#scope)
* [Inputs](#inputs)
* [Summary](#summary)
  * [Restrictions on combinations of the message tags](#restrictions-on-combinations-of-the-message-tags)
* [Test procedure]
* [Outcome(s)](#outcomes)
* [Special procedural requirements](#special-procedural-requirements)
* [Intercase dependencies](#intercase-dependencies)
* [Terminology](#terminology)

## Objective

DNSSEC is the security upgrade of DNS, just as TLS is the security upgrade of
HTTP (but done in very different ways). If a zone is upgrades to DNSSEC it means
that it is signed by DNSSEC keys. This test case will verify if the zone has been
DNSSEC signed, and if so, also verify that there is at least one DS record in the
parent zone for the zone tested.

The public half of the DNSSEC keys are stored in the zone. For a zone to be
correctly signed it is not enough to have DNSKEY records, but this test case
assumes that the existence of at least one DNSKEY record with at least one RRSIG
record covering the DNSKEY RRset means that the zone is signed.

> It is not enough to have a signed zone. The parent zone must have DS record to

Reviewer: ```suggestion
It is not enough to have a signed zone. The parent zone must have a DS record to
```

create a chain of trust from root. If the zone is signed (has at least one
DNSKEY record) this test case will check if the parent zone has any DS records
for the zone.

The method for authentication a DNS response is described in section 5 of
[RFC 4035][RFC 4035#section-5]. The DNSKEY record is defined in section 2 of
[RFC 4034][RFC 4034#section-2], and the DS record is defined in section 5 of
[RFC 4034][RFC 4034#section-5].

> Signing the zone and providing DS records for a signed zone is not reuired by

Reviewer: ```suggestion
Signing the zone and providing DS records for a signed zone is not required by
```

the protocol, but it is a deviation from best practices that should trigger an

> an alert. A WARNING is therefore raised if the zone is not signed, or if DS

Reviewer: ```suggestion
alert. A WARNING is therefore raised if the zone is not signed, or if DS
```

| DS07_NOT_SIGNED               | WARNING |                | The child zone is not signed.                                                                                                              |
| DS07_NOT_SIGNED_ON_SERVER     | WARNING | ns_list        | The following name servers responds with no DNSKEY (unsigned child zone). Servers: {ns_list}.                                              |
| DS07_NO_DS_ON_PARENT_SERVER   | WARNING | ns_list        | The following parent name servers responds without DS record for the child zone. Servers: {ns_list}.                                        |
| DS07_NO_DS_FOR_SIGNED_ZONE    | WARNING |                | The parent zone has no DS record for the signed child zone.                                                                                |
| DS07_NO_RESPONSE_DNSKEY       | WARNING | ns_list        | The following name servers dig not respond on DNSKEY query with DO bit set. Servers: {ns_list}                                             |

> | DS07_SIGNED                   | INFO    |                | The child zone is signed.                                                                                                                  |

Reviewer: Unnecessary here too I think.
```suggestion
| DS07_SIGNED                   | INFO    |                | The zone is signed.                                                                                                                        |
```

> | DS07_SIGNED_ON_SERVER         | INFO    | ns_list        | The following name servers responds with DNSKEY (signed child zone). Servers: {ns_list}.                                                   |

Reviewer: ```suggestion
| DS07_SIGNED_ON_SERVER         | INFO    | ns_list        | The following name servers responds with DNSKEY (signed child zone). Name servers: {ns_list}.                                              |
```

> | DS07_UNEXP_RCODE_RESP_DNSKEY  | WARNING | ns_list, rcode | The following name servers responed with RCODE "{rcode}" instead of expected "NOERROR" on DNSKEY query with DO bit set. Servers: {ns_list} |

Reviewer: ```suggestion
| DS07_UNEXP_RCODE_RESP_DNSKEY  | WARNING | ns_list, rcode | The following name servers responded with RCODE "{rcode}" instead of expected "NOERROR" on DNSKEY query with DO bit set. Name servers: {ns_list} |
```

The value in the Level column is the default severity level of the message. The
severity level can be changed in the [Zonemaster-Engine profile]. Also see the
[Severity Level Definitions] document.

The argument names in the Arguments column lists the arguments used in the
message. The argument names are defined in the [argument list].

The name server names are assumed to be available at the time when the msgid
is created, if the argument name is "ns" or "ns_list" even when in the
"[Test procedure]" below it is only referred to the IP address of the name
servers.

### Restrictions on combinations of the message tags

Below are some restrictions on how some of the message tags, defined above, can
be combined.

> * For each test exact one of *DS07_INCONSISTENT_SIGNED*, *DS07_SIGNED* and

Reviewer: ```suggestion
* For each test exactly one of either *DS07_INCONSISTENT_SIGNED*, *DS07_SIGNED* and
```

6.  For each name server IP address in *Child NS IP* do:

    1.  Send *SOA Query* to the name server IP.
    2.  Add the name server IP to the *Ignored Child NS* set and go to next name
        server IP if at least one of the following criteria is met:
          1. There is no DNS response.
          2. The [RCODE Name] in the response is not "NoError".
          3. The AA flag is not set in the response.
          4. The SOA record is missing in the answer section.
    3.  Send *DNSKEY Query* to the name server IP.
    4.  Add the name server IP to the *No Response DNSKEY Query* set and go to
        next name server IP if there is no DNS response to the query.
    5.  Add the name server IP to the *No Auth DNSKEY Response* set and go to
        next name server IP if the AA bit is not set in the response.
    6.  Add the name server IP and the [RCODE Name] to the
        *Error RCODE DNSKEY Response* set and go to next name server IP if the
        [RCODE Name] is not NOERROR.
    7.  Add the name server IP to the "Signed Response" set and go to next name
        server IP if the answer section contains the following DNS records:
        * At least one DNSKEY record.

>         * At least one RRSIG record covering the DNSKEY RRset.

Author: Clarified.

          done in this test case).
    8.  Else add the name server IP to the *No DNSKEY* set and go to
        next name server IP.

7.  Retrieve all name server IP addresses for the parent of *Child Zone* using
    method [Get-Parent-NS-IP] ("Parent NS IP").

8.  If *Undelegated DS* is non-empty then do:

    1. Add name server IP as "-" to the *DS in Response* set.
    2. Make *Parent NS IP* an empty set.

9.  If the *Signed Response* set is empty, make *Parent NS IP* and
    *DS in Response* empty sets.

>   Note: *Parent NS IP* will be empty if any of the followng is true:
>     * *Undelegated test* is TRUE.
>     * *Undelegated DS* is non-empty.
>     * *Child Zone* is ".", i.e. root zone.

> >     * The *Signed Response* set is empty.

Reviewer 2: The bullet point list does not render properly on GitHub:

<img width="997" height="65" alt="image" src="https://github.com/user-attachments/assets/59136d46-5a6b-4d8b-af7a-adb243d94483" />

10. For each parent name server IP in *Parent NS IP* do:
    1. Send *DS Query* to the name server IP.
    2. If at least one of the following criteria is met, then add name server IP
       to the "Ignored Parent NS IP" set and go to next parent name server:
       1. There is no [DNSSEC Response].
       2. The RCODE in the [DNSSEC Response] is not "NoError"
          ([IANA RCODE List]).
       3. The OPT record is absent in the [DNSSEC Response].
       4. The DO flag is unset in the [DNSSEC Response].
       5. The AA flag is not set in the [DNSSEC Response].
    3. Add the parent name server IP to the *DS in Response* set and go to the
       next parent name server IP if the answer section contains the following

>        DNS records:

Author: Yes, a DS without RRSIG is not enough.

       DNS records:
       * At least one DS record with *Child Zone* as owner name.

>        * At least one RRSIG covering the DS RRset.

Author: Clarified.

    4. Else add the parent name server IP to the *No DS* set and go
       to next parent name server IP.

11. If the following sets combined is identical to the *Child NS IP* set, output

>     *[DS07_NOT_SIGNED]*.

Author: Clarified.

    6.  Add the name server IP and the [RCODE Name] to the
        *Error RCODE DNSKEY Response* set and go to next name server IP if the
        [RCODE Name] is not NOERROR.
    7.  Add the name server IP to the *Signed Response* set and go to next name
        server IP if the answer section contains the following DNS records:
        * At least one DNSKEY record.
        * At least one RRSIG record [covering] the DNSKEY RRset.
    8.  Else add the name server IP to the *No DNSKEY* set and go to
        next name server IP.

7.  Retrieve all name server IP addresses for the parent of *Child Zone* using
    method [Get-Parent-NS-IP] ("Parent NS IP").

8.  If *Undelegated DS* is non-empty then do:

    1. Add name server IP as "-" to the *DS in Response* set.
    2. Make *Parent NS IP* an empty set.

9.  If the *Signed Response* set is empty, make *Parent NS IP* and
    *DS in Response* empty sets.

> Note: *Parent NS IP* will be empty if any of the followng is true:
> * *Undelegated test* is TRUE.
> * *Undelegated DS* is non-empty.
> * *Child Zone* is ".", i.e. root zone.
> * The *Signed Response* set is empty.

> 10. For each parent name server IP in *Parent NS IP* do:

Reviewer: ```suggestion
10. For each unique parent name server IP in *Parent NS Names and IPs* do:
```

9.  If the *Signed Response* set is empty, make *Parent NS IP* and
    *DS in Response* empty sets.

> > Note: *Parent NS IP* will be empty if any of the followng is true:

Reviewer 2: ```suggestion
> Note: *Parent NS IP* will be empty if any of the following is true:
```

[DS07_UNEXP_RCODE_RESP_DNSKEY]:               #summary
[ERROR]:                                      ../SeverityLevelDefinitions.md#error
[Get-Del-NS-Names-and-IPs]:                   ../MethodsV2.md#method-get-delegation-ns-names-and-ip-addresses
[Get-Parent-NS-IP]:                           ../MethodsV2.md#method-get-parent-ns-ip-addresses
[Get-Zone-NS-Names-and-IPs]:                  ../MethodsV2.md#method-get-zone-ns-names-and-ip-addresses
[IANA RR Type List]:                          https://www.iana.org/assignments/dns-parameters/dns-parameters.xhtml#dns-parameters-4
[INFO]:                                       ../SeverityLevelDefinitions.md#info
[NOTICE]:                                     ../SeverityLevelDefinitions.md#notice
[RCODE Name]:                                 https://www.iana.org/assignments/dns-parameters/dns-parameters.xhtml#dns-parameters-6
[RFC 4034#section-2]:                         https://datatracker.ietf.org/doc/html/rfc4034#section-2
[RFC 4034#section-5]:                         https://datatracker.ietf.org/doc/html/rfc4034#section-5

> [RFC 4035#section-5.3.1]:                     https://datatracker.ietf.org/doc/html/rfc4035#section-5.3.1]

Reviewer 2: ```suggestion
[RFC 4035#section-5.3.1]:                     https://datatracker.ietf.org/doc/html/rfc4035#section-5.3.1
```
